KEP-4004: Deprecate the kubeProxyVersion field of v1.Node (#4005)
kubernetes/enhancements: keps/sig-node/4004-deprecate-kube-proxy-version/README.md

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Deprecate the `status.nodeInfo.kubeProxyVersion` field of v1.Node

## Motivation

This field is not accurate, the field is set by kubelet, which does not actually know the kube-proxy version,
or even if kube-proxy is running.

### Goals

- Mark `status.nodeInfo.kubeProxyVersion` deprecated.
- Find any places use it and get them to stop.
- Make kubelet stop setting the field.

### Non-Goals

## Proposal

> - The proposal is to deprecate the kubeProxyVersion field on NodeStatus, which would involve modifying the Kubernetes API and would affect GCP cloud providers (e.g. https://github.com/kubernetes/cloud-provider-gcp/pull/533, https://github.com/kubernetes/kubernetes/pull/117806) or other components, so we will use a feature gates to protect it until we fix the GCP cloud provider and any other components we find using the deprecated fields

Reviewer: ```suggestion
- The proposal is to deprecate the kubeProxyVersion field on NodeStatus, which would involve modifying the Kubernetes API and would affect GCP cloud providers (e.g. https://github.com/kubernetes/cloud-provider-gcp/pull/533, https://github.com/kubernetes/kubernetes/pull/117806) or other components, so we will use a feature gates to protect it until we fix the GCP cloud provider and any other components we find using the deprecated fields
```

can we be more explicit? what do we mean by "modifying the Kubernetes API " 
 I thought that we were going to stop setting the field only and set as "deprecated" on the generated API docs

Author: > can we be more explicit? what do we mean by "modifying the Kubernetes API " I thought that we were going to stop setting the field only and set as "deprecated" on the generated API docs

Sorry, setting the field to deprecated shouldn't count as an API change, I'll change the wording

### Risks and Mitigations

## Design Details

### Versioned API Change: NodeStatus v1 core

Mark the kubeProxyVersion field as Deprecated.

```
// Deprecated: Kubelet Version reported by the node.

> KubeletVersion string `json:"kubeletVersion" protobuf:"bytes,7,opt,name=kubeletVersion"`

Reviewer 2: Why does this say Kubelet ?

Author: Oh, sorry, that was a stupid mistake

```
// Deprecated: KubeProxy Version reported by the node.
KubeProxyVersion string
```

### Test Plan

[x ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- Add a test to `TestVersionInfo` in `pkg/kubelet/nodestatus` to see if FeatureGate behaves as expected when it is turned on or off.
- If DisableNodeKubeProxyVersion FeatureGate is enabled:
  - `status.nodeInfo.kubeProxyVersion` will be empty.
- Else:
  - `status.nodeInfo.kubeProxyVersion` will be not empty .

##### Integration tests

- N/A

##### e2e tests

 - N/A

### Graduation Criteria

1. Alpha: v1.28, Created the feature gate, disabled by default, and started looking for components that might be using the deprecated field.
2. Beta: v1.29, Make sure that we fixed the GCP cloud provider and any other components we found that use the deprecated field, and we need to make sure that upgrades to the Beta version will work across supported levels of [version skew](https://kubernetes.io/releases/version-skew-policy/).

> 3. GA: v1.30

Author: > These should be separate sections; see the example in the template README

all right

> keps/sig-node/4004-deprecate-kube-proxy-version/kep.yaml

Author: yes...

  - Components depending on the feature gate: kubelet

###### Does enabling the feature change any default behavior?

Yes, it makes kubelet stop setting `node.status.nodeInfo.kubeProxyVersion`.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Using the featuregate is the only way to enable/disable this feature.

###### What happens if we reenable the feature if it was previously rolled back?

The feature should continue to work just fine.

###### Are there any tests for feature enablement/disablement?

No, we don't need to test what happens when the value of the field changes, because our goal is to make sure no one is looking at the value at all, so no components should notice if the value changes.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

If there is a component that we didn't know about using kubeProxyVersion, then it may fail in some unknown way when the administrator upgrades to a version with that field disabled. It should be possible to just roll back in this case.

###### What specific metrics should inform a rollback?

> No

Reviewer 3: ```suggestion
N/A
```

Author: fixed

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> No

Reviewer 3: ```suggestion
N/A
```

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Deprecate the `status.nodeInfo.kubeProxyVersion` field of v1.Node

## Motivation

> This field is not accurate, the field is set by kubelet, which does not actually know the kube-proxy version

Reviewer 3: ```suggestion
This field is not accurate, the field is set by kubelet, which does not actually know the kube-proxy version,
or even if kube-proxy is running.
```

### Goals

> Deprecate the `status.nodeInfo.kubeProxyVersion` field and then do the hard work of finding all the places using it and getting them off in time.

Reviewer 3: So I'd phrase that as:

```
- Mark `status.nodeInfo.kubeProxyVersion` deprecated.
- Find any places use it and get them to stop.
- Make kubelet stop setting the field.
```

### Non-Goals

## Proposal

- Deprecate the `status.nodeInfo.kubeProxyVersion` field of v1.Node

> - Remove the gcp cloud provider check for this field

Reviewer 3: ah, yeah, I feel like this is more "Goals" than "Proposal"...

Also, wherever it is that you mention the GCP cloud provider, you need to be clear that that's just one specific example, and that we may find others.

(Also, maybe link to the PRs I already filed: https://github.com/kubernetes/cloud-provider-gcp/pull/533, https://github.com/kubernetes/kubernetes/pull/117806.)

Author: > ah, yeah, I feel like this is more "Goals" than "Proposal"...
> 
> Also, wherever it is that you mention the GCP cloud provider, you need to be clear that that's just one specific example, and that we may find others.
> 
> (Also, maybe link to the PRs I already filed: [kubernetes/cloud-provider-gcp#533](https://github.com/kubernetes/cloud-provider-gcp/pull/533), [kubernetes/kubernetes#117806](https://github.com/kubernetes/kubernetes/pull/117806).)

I have made some changes, please see if it is more appropriate

> ### User Stories (Optional)

Reviewer 3: You can remove any "(Optional)" sections you aren't using.

Given that this is a pretty minimal KEP (since it's not a "real" feature) you can remove any other normally-non-optional sections you aren't filling in too.

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

### Versioned API Change: NodeStatus v1 core

Mark the kubeProxyVersion field as Deprecated

```
//Deprecated: Kubelet Version reported by the node.
KubeletVersion string
```

### NodeStatus Internal Representation

Mark the kubeProxyVersion field as Deprecated

```
//Deprecated: Kubelet Version reported by the node.

> KubeletVersion string `json:"kubeletVersion" protobuf:"bytes,7,opt,name=kubeletVersion"`

Reviewer 3: pretty sure these two code snippets are reversed; it's the v1 version that has the json/protobuf annotations, and the internal one that doesn't.

also, space between `//` and `Deprecated`

```

### Test Plan

> [ ] I/we understand the owners of the involved components may require updates to

Reviewer 3: You're supposed to `[X]` this, but again, since this is a minimal KEP, you can just eliminate a lot of this; we will need to update the kubelet nodestatus unit tests to expect the new behavior rather than the old, but other than that, there's nothing we need to explicitly test. We just want to make sure that other tests keep passing.

existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Every code should have complete unit test coverage

##### Integration tests

- N/A

##### e2e tests

- Test whether FeatureGate behaves as expected when it is turned on or off
- If kubeProxyVersion FeatureGate is enabled:
  - `status.nodeInfo.kubeProxyVersion` will be empty
- Else:

>   - `status.nodeInfo.kubeProxyVersion` will be not empty 

Reviewer 3: This can be done in the unit tests in `pkg/kubelet/nodestatus`, which is a lot easier than testing feature-gate-related stuff in e2e tests.

> ### Graduation Criteria

Reviewer 3: We need this section filled in.

For Alpha, we should have created the feature gate, disabled by default, and started looking for components that might be using the deprecated field.

For Beta, we want to make sure that we fixed the GCP cloud provider and any other components we found that use the deprecated field, and we need to make sure that upgrades to the Beta version will work across supported levels of [version skew](https://kubernetes.io/releases/version-skew-policy/). eg, if there are components in 1.28 that depend on `kubeProxyVersion` being set, and it's allowable for a cluster to have a 1.29 kubelet and a 1.28 version of the component that depends on `kubeProxyVersion`, then we can't safely go to beta in 1.29.

### Upgrade / Downgrade Strategy

> ### Version Skew Strategy

Reviewer 3: Upgrade/downgrade is basically "N/A". For skew, see what I said above.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

> - [ ] Feature gate (also fill in values in `kep.yaml`)

Reviewer 3: `[X]` this, and delete the `[ ] Other` section

  - Feature gate name: kubeProxyVersion
  - Components depending on the feature gate: kubelet
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?

> ###### Does enabling the feature change any default behavior?

Reviewer 3: Yes; it makes kubelet stop setting `node.status.nodeInfo.kubeProxyVersion`.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Using the featuregate is the only way to enable/disable this feature.

###### What happens if we reenable the feature if it was previously rolled back?

The feature should continue to work just fine.

###### Are there any tests for feature enablement/disablement?

> No, these will be introduced in the Alpha phase.

Reviewer 3: That would be "Yes", not "No" because you're supposed to include the work being done for Alpha in the answers here.

But I think this should be "No" anyway; "tests for feature enablement/disablement" means "tests where you have a running cluster with the feature disabled and then you enable it", or vice versa. And there's no reason to do that here; we need to make sure no one is looking at the deprecated field ever, so we don't need to test what happens when the value of the field _changes_, since if anyone was actually looking at it all, then we're already failing.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Deprecate the `status.nodeInfo.kubeProxyVersion` field of v1.Node

## Motivation

This field is not accurate, the field is set by kubelet, which does not actually know the kube-proxy version,
or even if kube-proxy is running.

### Goals

- Mark `status.nodeInfo.kubeProxyVersion` deprecated.
- Find any places use it and get them to stop.
- Make kubelet stop setting the field.

### Non-Goals

## Proposal

> - The proposal is to deprecate the kubeProxyVersion field on NodeStatus, which would affect GCP cloud providers (e.g. https://github.com/kubernetes/cloud-provider-gcp/pull/533, https://github.com/kubernetes/kubernetes/pull/117806) or other components, so we will use a feature gates to protect it until we fix the GCP cloud provider and any other components we find using the deprecated fields

Reviewer 3: At this point, both of those PRs have merged; the cloud-provider-gcp fix will be in 1.29. So...

```suggestion
The proposal is to deprecate the `kubeProxyVersion` field of `NodeStatus`, and to stop setting it in the future.

This field was used by the GCP cloud provider up until 1.28 for the legacy built-in cloud provider ([kubernetes #117806], and up until 1.29 for the external cloud-provider ([cloud-provider-gcp #533]). It may also be used by other components. Thus, we will use a feature gate to protect it until all components are fixed.

[kubernetes #117806]: https://github.com/kubernetes/kubernetes/pull/117806
[cloud-provider-gcp #533]: https://github.com/kubernetes/cloud-provider-gcp/pull/533
```

Reviewer 3: actually, those versions are off by one; the bug was _fixed_ in cloud-provider-gcp as of 1.29, so that means it used the field up until 1.28. And likewise 1.27 for the legacy provider.

Author: All right

### Risks and Mitigations

## Design Details

### Versioned API Change: NodeStatus v1 core

Mark the kubeProxyVersion field as Deprecated.

```
// Deprecated: KubeProxy Version reported by the node.
KubeProxyVersion string `json:"kubeProxyVersion" protobuf:"bytes,8,opt,name=kubeProxyVersion"`
```

### NodeStatus Internal Representation

Mark the kubeProxyVersion field as Deprecated.

```
// Deprecated: KubeProxy Version reported by the node.
KubeProxyVersion string
```

### Test Plan

[x ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- Add a test to `TestVersionInfo` in `pkg/kubelet/nodestatus` to see if FeatureGate behaves as expected when it is turned on or off.
- If DisableNodeKubeProxyVersion FeatureGate is enabled:
  - `status.nodeInfo.kubeProxyVersion` will be empty.
- Else:
  - `status.nodeInfo.kubeProxyVersion` will be not empty .

> ##### Integration tests
> 
> 
> - N/A

Reviewer 3: ```suggestion
##### Integration tests

- N/A
```

##### e2e tests

 - N/A

### Graduation Criteria

#### Alpha

> - [ ] Created the feature gate, disabled by default.

Reviewer 3: oh, we don't normally bother with checkboxes here...

Author: fixed

- [ ] Started looking for components that might be using the deprecated field.

#### Beta

> - [ ] Make sure that we fixed the GCP cloud provider and any other components we found that use the deprecated field.

Reviewer 3: GCP cloud provider is already done at this point so that doesn't need to be mentioned here

Author: removed

N/A

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: DisableNodeKubeProxyVersion
  - Components depending on the feature gate: kubelet

###### Does enabling the feature change any default behavior?

Yes, it makes kubelet stop setting `node.status.nodeInfo.kubeProxyVersion`.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Using the featuregate is the only way to enable/disable this feature.

###### What happens if we reenable the feature if it was previously rolled back?

The feature should continue to work just fine.

###### Are there any tests for feature enablement/disablement?

No, we don't need to test what happens when the value of the field changes, because our goal is to make sure no one is looking at the value at all, so no components should notice if the value changes.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> 

Reviewer 3: If there is a component that we didn't know about using `kubeProxyVersion`, then it may fail in some unknown way when the administrator upgrades to a version with that field disabled. It should be possible to just roll back in this case.

###### What specific metrics should inform a rollback?

> ###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Reviewer 3: answer these questions with "N/A" or "No", as appropriate; when they're empty it looks like you just haven't gotten around to filling in this section of the KEP yet

Author: fixed

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

> ### Monitoring Requirements

Reviewer 3: You can just put "N/A" here and delete the sub-sections of this section

Author: fixed
